utils/writeable/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+//! `writeable` is a utility crate of the [`ICU4X`] project.
+//!
+//! It includes [`Writeable`], a core trait representing an object that can be written to a
+//! sink implementing std::fmt::Write. It is an alternative to std::fmt::Display with the
+//! addition of a function indicating the number of bytes to be written.
+//!
+//! Writeable improves upon std::fmt::Display in two ways:
+//!
+//! 1. More efficient, since the sink can pre-allocate bytes.
+//! 2. Smaller code, since the format machinery can be short-circuited.
+//!
+//! Types implementing Writeable automatically implement ToString. Because of this, you cannot
+//! implement both Writeable and std::fmt::Display on the same type.
+//!
+//! [`Writeable`]: ./trait.Writeable.html
+//! [`ICU4X`]: ../icu/index.html
+
+use std::fmt;
+use std::string::ToString;
+
+/// Writeable is an alternative to std::fmt::Display with the addition of a length function.
+pub trait Writeable {
+    /// Writes bytes to the given sink. Errors from the sink are bubbled up.
+    fn write_to(&self, sink: &mut dyn fmt::Write) -> fmt::Result;
+
+    /// Returns an estimation of the number of bytes that will be written to the sink. The actual
+    /// number of bytes may be slightly different than what this function returns.
+    ///
+    /// This function may return an enumeration in the future. See:
+    /// https://github.com/unicode-org/icu4x/issues/370
+    fn write_len(&self) -> usize;
+}
+
+impl ToString for dyn Writeable {
```

Comment:
This is useful but as you note we need to explicitly cast to `dyn Writeable`, which isn't great.

What if we provide a defaulted `.writeable_to_string()` method on `Writeable` that does this, and if you want the fast ToString you call `.writeable_to_string()` instead?